File: vijaymamra/public/py/payment_entry.py
import frappe


@frappe.whitelist()
def check_entries(payment_entry_name=None):
    payment_entries = frappe.get_all("Payment Entry", filters={"status": "Submitted"}, fields=["name"])
    
    bank_transactions = frappe.get_all("Bank Transaction", filters={"status": "reconciled"}, fields=["name"])

    matched_entries = []

    for bank_transaction in bank_transactions:
        bank_transaction_doc = frappe.get_doc("Bank Transaction", bank_transaction['name'])
        for payment in bank_transaction_doc.payment_entries:
            for payment_entry in payment_entries:
                if payment.payment_entry == payment_entry['name']:
                    matched_entries.append(payment_entry['name'])

    if payment_entry_name and payment_entry_name in matched_entries:
        return {"matched": True, "entries": matched_entries}
    return {"matched": False, "entries": matched_entries}

File: vijaymamra/public/js/payment_entry.js
frappe.ui.form.on('Payment Entry', {
    before_cancel: function(frm) {
        frappe.call({
            method: 'vijaymamra.public.py.payment_entry.check_entries',
            args: {
                payment_entry_name: frm.doc.name
            },
            callback: function(response) {
                console.log(response)
                if (response.message.matched) {
                    frappe.msgprint(__('This Payment Entry is linked with a reconciled Bank Transaction and cannot be canceled.'));
                    frappe.validated = false;
                } else {
                    frm.savesubmit(); 
                }
            }
        });
    }
});
// apps/vijaymamra/vijaymamra/public/py/payment_entry.py